### PR TITLE
Added rolling hash calculation in deli

### PR DESCRIPTION
**Pull request by tanviraumi**
Following up on #5689

This PR adds rolling hash calculation in deli.

Open questions?

1. We are storing the hash as a string since it's more generic. However, the calculation generates a number which means we have to parse it back and forth for each calculation. Wondering whether we should just use a number as the hash field.
2. We are using JSON.stringify() to generate string based on a set of fields before we hash. Alternatively, we can use our own serialization format to generate a more compact and unique string.
3. Using https://www.npmjs.com/package/crc-32 to calculate the hash for now. Should we invest on writing our own hashing algorithm?
4. This PR only addresses deli. In future, we need to add the hash in summary as well.

### CODE REVIEW

**Comment by tanviraumi** [comment]
@vladsud and @anthony-murphy: Please take a look. The goal is to evolve this PR and implement something we can agree upon

File: server/routerlicious/packages/lambdas/src/deli/lambda.ts [hunk 1]
```diff
@@ -181,6 +182,7 @@ export class DeliLambda extends TypedEventEmitter<IDeliLambdaEvents> implements
 
         // Initialize counting context
         this.sequenceNumber = lastCheckpoint.sequenceNumber;
+        this.lastHash = lastCheckpoint.expHash1 ?? "00000000";
```
Comment (GaryWilber) [inline server/routerlicious/packages/lambdas/src/deli/lambda.ts:185]: Can this `"00000000"` be moved to the rollingHash file as a constant

File: server/routerlicious/packages/services-client/src/rollingHash.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    const sequenceNumberProperty: keyof ISequencedDocumentSystemMessage = "sequenceNumber";
+    const minimumSequenceNumberProperty: keyof ISequencedDocumentSystemMessage = "minimumSequenceNumber";
+    const clientSequenceNumberProperty: keyof ISequencedDocumentSystemMessage = "clientSequenceNumber";
+    const referenceSequenceNumberProperty: keyof ISequencedDocumentSystemMessage = "referenceSequenceNumber";
+    const typeProperty: keyof ISequencedDocumentSystemMessage = "type";
+    const timestampProperty: keyof ISequencedDocumentSystemMessage = "timestamp";
+    const dataProperty: keyof ISequencedDocumentSystemMessage = "data";
+
+    // Ordering of these fields decides the ordering of serialized message.
+    const fields: string[] = [
```
Comment (GaryWilber) [inline server/routerlicious/packages/services-client/src/rollingHash.ts:22]: All these fields and properties be moved outside this method so it won't create the object/strings every time.

File: server/routerlicious/packages/services-client/src/rollingHash.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        minimumSequenceNumberProperty,
+        clientSequenceNumberProperty,
+        referenceSequenceNumberProperty,
+        typeProperty,
+        timestampProperty,
+        dataProperty,
+    ];
+
+    // Should we just use stringified JSON or come up with our own serialization of values?
+    const serializedMessage = JSON.stringify(messageWithData, fields);
```
Comment (GaryWilber) [inline server/routerlicious/packages/services-client/src/rollingHash.ts:34]: The [MDN docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify) say that `Properties of non-array objects are not guaranteed to be stringified in any particular order. Do not rely on ordering of properties within the same object within the stringification.`

If the order is not guaranteed, then maybe the hash could change even though it's the same op?

File: server/routerlicious/packages/services-client/package.json [hunk 4]
```diff
@@ -56,6 +56,7 @@
     "@fluidframework/protocol-base": "^0.1030.0",
     "@fluidframework/protocol-definitions": "^0.1025.0-0",
     "axios": "^0.21.1",
+    "crc-32": "1.2.0",
```
Comment (GaryWilber) [inline server/routerlicious/packages/services-client/package.json:59]: I know the the hash function is still a work in progress, but at some point we may want to use a different module for hashing in node.js
For example http://engineering.voxer.com/2013/12/02/hardwareaccelerated-crc32-for-nodejs/ / https://github.com/Voxer/sse4_crc32 uses a native node module for faster crc32's. Just something to keep in mind

File: server/routerlicious/packages/services-client/src/rollingHash.ts (hunk 2 above)
Comment (tanviraumi) [reply server/routerlicious/packages/services-client/src/rollingHash.ts:22]: good point. Will do!

File: server/routerlicious/packages/services-client/src/rollingHash.ts (hunk 3 above)
Comment (tanviraumi) [reply server/routerlicious/packages/services-client/src/rollingHash.ts:34]: Based on my understanding, the fields will determine the order (https://stackoverflow.com/questions/16167581/sort-object-properties-and-json-stringify). And our fields are primitive only which means we can use them easily.

File: server/routerlicious/packages/services-client/package.json (hunk 4 above)
Comment (tanviraumi) [reply server/routerlicious/packages/services-client/package.json:59]: Yes this one is just to get the ball rolling. Let me know if you prefer any specific one

**Review by anthony-murphy — APPROVED** [review]
looks like a good start. lets put this behind a flag, or config, and then get it in so we can start playing with it

**Review by GaryWilber — APPROVED** [review]

**Comment by tanviraumi** [comment]
Thanks @GaryWilber

**Pull request merged**